Contributing: Update guide for clang-format

leoetlino opened the pull request:
This changes the guide to reflect the new formatting rules, and adds
instructions for using clang-format to automatically check and fix
most formatting issues.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/dolphin-emu/dolphin/3952)
<!-- Reviewable:end -->

BhaaLseN commented on Contributing.md:
```diff
@@ -24,33 +24,55 @@ This guide is for developers who wish to contribute to the Dolphin codebase. It
+    m_other->DoStuff(m_first_parameter,
+                     m_second_parameter,
+                     m_third_parameter);
+    // Indent for the three previous lines is done using four tabs
```
This should be spaces too. The number 17 down below isn't correct either.
The question is how useful this example is, since `clang-format` probably forced it to have the parameters on the same line (given that they don't exceed the 100 character column). It used to serve as example for the mixed space/tab indent.

JosJuice commented on Contributing.md:
```diff
@@ -24,33 +24,55 @@ This guide is for developers who wish to contribute to the Dolphin codebase. It
+    // with four spaces.
+    m_other->DoStuff(m_first_parameter,
+                     m_second_parameter,
+                     m_third_parameter);
```
Please make a note to only do this kind of wrapping when it's necessary because the line would be too long otherwise. Doing what this example actually shows wouldn't be appropriate.

Alternatively (and probably preferably), you could remove the entire example. It was added to better explain the combination of tabs and spaces that we used before, because people got a bit confused when we only had text without an example. "Two spaces" is much easier to explain.

JosJuice commented on Contributing.md:
```diff
@@ -24,33 +24,55 @@ This guide is for developers who wish to contribute to the Dolphin codebase. It
+                     m_third_parameter);
+    // Indent for the three previous lines is done using four tabs
+    // each (which brings the lines to the column where the word
+    // m_other begins in the first line).
```
Four tabs? Don't you mean four spaces?

delroth commented on Contributing.md:
````diff
@@ -24,33 +24,55 @@ This guide is for developers who wish to contribute to the Dolphin codebase. It
+
+- To run clang-format on all staged files:
+  ```
+  git diff --cached --name-only | egrep '[.](cpp|h|mm)$' | xargs clang-format -style=file -i
````
-style=file shouldn't be necessary.

delroth commented on Contributing.md:
````diff
@@ -24,35 +24,36 @@ This guide is for developers who wish to contribute to the Dolphin codebase. It
+  ```
+  ln -sf ../../Tools/reformat.sh .git/hooks/pre-commit
+  ```
+  Note that the hook will completely reformat all staged files and commit them, so avoid using git commit -p or add -p or similar when using this hook.
````
Note that clang-format supports --lines / --offset to partially reformat a file. Maybe this could be used to only reformat the index.

leoetlino replied:
This is probably possible, but seems to be a bit complicated, as we would need to figure out the lines to add back to the index after reformatting. So I'll leave this as a future possible improvement :)

IMO, the proper way of doing it would be a git filter driver, where all changes are always just "cleaned" automatically and transparently, so git only ever sees the formatted files in the repo. This would probably solve the partial reformat issue as well.
Should I mention this along with, or potentially instead of the reformat hook?

leoetlino commented:
@dolphin-emu-bot rebuild

leoetlino commented:
Review status: 0 of 1 files reviewed at latest revision, 4 unresolved discussions.

---

*[Contributing.md, line 36 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3952#-KLCzUln0tK7gcNkdefN-r1-36:-KNqcPfDYSKIkqmNB78E:-499146009) ([raw file](https://github.com/dolphin-emu/dolphin/blob/a34ed0e8099edd9858e9b67b9a44cf622abe4cc2/Contributing.md#L36)):*
<details><summary><i>Previously, delroth (Pierre Bourdon) wrote…</i></summary>
> -style=file shouldn't be necessary.
</details>
Done.

---

*[Contributing.md, line 68 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3952#-KLCzUln0tK7gcNkdefN-r1-68:-KNqcSCBoOkrC3OuDEWt:-499146009) ([raw file](https://github.com/dolphin-emu/dolphin/blob/a34ed0e8099edd9858e9b67b9a44cf622abe4cc2/Contributing.md#L68)):*
<details><summary><i>Previously, JosJuice wrote…</i></summary>
> Please make a note to only do this kind of wrapping when it's necessary because the line would be too long otherwise. Doing what this example actually shows wouldn't be appropriate.
> 
> 
> 
> Alternatively (and probably preferably), you could remove the entire example. It was added to better explain the combination of tabs and spaces that we used before, because people got a bit confused when we only had text without an example. "Two spaces" is much easier to explain.
</details>
Done.

---

*[Contributing.md, line 71 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3952#-KLCzUln0tK7gcNkdefN-r1-71:-KNqci-oa8ICdz0XGh-0:1741885381) ([raw file](https://github.com/dolphin-emu/dolphin/blob/a34ed0e8099edd9858e9b67b9a44cf622abe4cc2/Contributing.md#L71)):*
<details><summary><i>Previously, JosJuice wrote…</i></summary>
> Four tabs? Don't you mean four spaces?
</details>
This was a mistake. Removed this along with the example.

---

*[Contributing.md, line 48 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3952#-KLCzUln0tK7gcNkdefN-r2-48:-KNqcq2Y8Lx3gV7Zbw2b:2031530299) ([raw file](https://github.com/dolphin-emu/dolphin/blob/cb2c16dc6e40211e278a5980128806f87c386669/Contributing.md#L48)):*
<details><summary><i>Previously, leoetlino (Léo Lam) wrote…</i></summary>
> This is probably possible, but seems to be a bit complicated, as we would need to figure out the lines to add back to the index after reformatting. So I'll leave this as a future possible improvement :)
> 
> 
> 
> IMO, the proper way of doing it would be a git filter driver, where all changes are always just "cleaned" automatically and transparently, so git only ever sees the formatted files in the repo. This would probably solve the partial reformat issue as well.
> 
> Should I mention this along with, or potentially instead of the reformat hook?
</details>
Changed this to suggest a git filter driver.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3952)*
<!-- Sent from Reviewable.io -->
